### PR TITLE
Allow populate block logic in compact to be overriden outside Prometheus

### PR DESCRIPTION
**Pull request by alexqyle**
Introduced `PopulateBlockFunc` interface so that the populate block logic in compact can be overriden outside Prometheus. For example, Thanos could add parallelism when populating blocks during compaction.

 Used mlock for Unix system when reading index to prevent swap.

Signed-off-by: Alex Le <leqiyue@amazon.com>

### CODE REVIEW

File: tsdb/index/postings.go [hunk 1]
```diff
@@ -409,6 +410,10 @@ type Postings interface {
 	Err() error
 }
 
+type AdditionalPostingsProvider interface {
```
Comment (fpetkovski) [inline tsdb/index/postings.go:413]: I would suggest renaming this to `PartitionPostingsFunc`, to be in line with `VerticalChunkSeriesMergeFunc`.

**Review by fpetkovski — COMMENTED** [review]
Thanks for the work. I would recommend removing the parallelism for now because it's slightly out of scope of this change. It's also maybe something that should be made configurable, right now it's unbounded.

**Comment by alexqyle** [comment]
> Thanks for the work. I would recommend removing the parallelism for now because it's slightly out of scope of this change. It's also maybe something that should be made configurable, right now it's unbounded.

Let me do some performance test with and without the parallelism to make sure doing extra sharded postings will not affect the performance that much.

**Comment by alexqyle** [comment]
> Thanks for the work. I would recommend removing the parallelism for now because it's slightly out of scope of this change. It's also maybe something that should be made configurable, right now it's unbounded.

@fpetkovski I updated the PR and did some performance test. Without the parallelism introduced in this PR, one compaction took one more hour to complete compared to with parallelism. The compaction time increased from 7 hours to 8 hours for 500M time series.

File: tsdb/compact.go [hunk 2]
```diff
@@ -808,3 +828,7 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 
 	return nil
 }
+
+type AdditionalPostingsFunc interface {
+	GetPostings(originalPostings index.Postings, indexReader IndexReader) index.Postings
```
Comment (yeya24) [inline tsdb/compact.go:833]: It would be nice to have some comments on the interface and method.

File: storage/merge.go [hunk 3]
```diff
@@ -333,10 +333,23 @@ func (c *genericMergeSeriesSet) Next() bool {
...
+				if c.currentSets[i].Next() {
+					ch <- c.currentSets[i]
+				}
+			}()
+		}
+
+		wg.Wait()
+		close(ch)
+		for set := range ch {
+			heap.Push(&c.heap, set)
```
Comment (yeya24) [inline storage/merge.go:352]: Can we make this change in a separate pr? And I think it might not be required to change it upstream if we really need this capability in Thanos/Cortex.
Can we just specify a different merge func here https://github.com/prometheus/prometheus/blob/main/tsdb/compact.go#L84?

File: tsdb/compact.go [hunk 4]
```diff
@@ -684,6 +691,8 @@ func (c *LeveledCompactor) populateBlock(blocks []BlockReader, meta *BlockMeta,
 	c.metrics.populatingBlocks.Set(1)
 
 	globalMaxt := blocks[0].Meta().MaxTime
+	g, _ := errgroup.WithContext(c.ctx)
+	g.SetLimit(8)
```
Comment (yeya24) [inline tsdb/compact.go:695]: Let's remove the concurrency change in this pr.
It makes sense for Cortex/Thanos to use more cores during compaction since they have dedicated compactor components. But Prometheus needs CPUs for different work like scrape, query, rules, TSDB, etc so this might not be something makes sense to Prometheus.

**Review by codesome — COMMENTED** [review]
While I am positive about this change, I suggested a different design. Compactor is used as an interface (where you compact with `Compact()` and `Write()` functions) and LeveledCompactor is just an implementation of it. It would be simpler to not require type inference to use this new method and keep using the interface. We can extend the `NewLeveledCompactor` to take an optional populateBlock function, and if provided, it will use that function instead of the default one. And while using it in your projects, you can have one instance of compactor per custom populateBlock function and continue to use the `Compact()` function.

File: tsdb/compact.go [hunk 5]
```diff
@@ -84,28 +84,28 @@ type LeveledCompactor struct {
...
+type CompactorMetrics struct {
 	ran               prometheus.Counter
-	populatingBlocks  prometheus.Gauge
-	overlappingBlocks prometheus.Counter
+	PopulatingBlocks  prometheus.Gauge
+	OverlappingBlocks prometheus.Counter
 	duration          prometheus.Histogram
 	chunkSize         prometheus.Histogram
 	chunkSamples      prometheus.Histogram
 	chunkRange        prometheus.Histogram
```
Comment (codesome) [inline tsdb/compact.go:94]: It makes sense to expose all the metrics at this point instead of a selected few

File: tsdb/compact.go [hunk 6]
```diff
@@ -659,10 +663,16 @@ func (c *LeveledCompactor) write(dest string, meta *BlockMeta, blocks ...BlockRe
 	return nil
 }
 
-// populateBlock fills the index and chunk writers with new data gathered as the union
+type PopulateBlockFunc interface {
+	PopulateBlock(ctx context.Context, metrics *CompactorMetrics, logger log.Logger, chunkPool chunkenc.Pool, mergeFunc storage.VerticalChunkSeriesMergeFunc, blocks []BlockReader, meta *BlockMeta, indexw IndexWriter, chunkw ChunkWriter) error
+}
```
Comment (codesome) [inline tsdb/compact.go:668]: How about just this? Is having it as an interface more useful?
```suggestion
type PopulateBlockFunc func(ctx context.Context, metrics *CompactorMetrics, logger log.Logger, chunkPool chunkenc.Pool, mergeFunc storage.VerticalChunkSeriesMergeFunc, blocks []BlockReader, meta *BlockMeta, indexw IndexWriter, chunkw ChunkWriter) error
```

**Comment by alexqyle** [comment]
> While I am positive about this change, I suggested a different design. Compactor is used as an interface (where you compact with `Compact()` and `Write()` functions) and LeveledCompactor is just an implementation of it. It would be simpler to not require type inference to use this new method and keep using the interface. We can extend the `NewLeveledCompactor` to take an optional populateBlock function, and if provided, it will use that function instead of the default one. And while using it in your projects, you can have one instance of compactor per custom populateBlock function and continue to use the `Compact()` function.

Here is Thanos side proposed PR to demostrate how Thanos would use this change from Prometheus: https://github.com/thanos-io/thanos/pull/5964/files#diff-be21b0381ac14ee139355d057b390aa69ff338fe1c748ae8a475ce46f3583575R1140

Thanos would create `PopulateBlockFunc` instance with additional information then pass it into `CompactWithPopulateBlockFunc` and Thanos has its own version of [Compactor interface](https://github.com/thanos-io/thanos/pull/5964/files#diff-be21b0381ac14ee139355d057b390aa69ff338fe1c748ae8a475ce46f3583575R800). If it is better to create a new implementation `NewLeveledCompactor`, I think it would be easy just keep the new implementation Thanos only without touching anything in Prometheus. This also means Thanos would have a copy of `LeveledCompactor` with additional logic on top of it. In order to avoid this duplication, I decided to do in this way.

File: tsdb/index/reader_builder_default.go [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+
+package index
+
+import (
+	"fmt"
+	"io"
+	"syscall"
+)
+
+func newReader(b ByteSlice, c io.Closer) (*Reader, error) {
```
[the remark this answers is not in the document]
Comment (alexqyle) [reply tsdb/index/reader_builder_default.go:25]: For sure, I can make a different PR for reader mlock change.

**Comment by alexqyle** [comment]
@codesome @fpetkovski Do you guys have any chance to take a look the code change and my comment? I need some inputs from you guys

File: tsdb/querier.go [hunk 8]
```diff
@@ -891,7 +891,7 @@ type mergedStringIter struct {
 }
 
 func (m *mergedStringIter) Next() bool {
-	if (!m.aok && !m.bok) || (m.Err() != nil) {
+	if !m.aok && !m.bok {
```
Comment (codesome) [inline tsdb/querier.go:894]: We should not ignore the error here. This change should be reverted. That said, https://github.com/prometheus/prometheus/pull/12132 made some changes here recently. Can you update your branch with latest changes in the main instead?

**Review by codesome — COMMENTED** [review]
The changes look good. Can you please fix the below comment and the ones posted before?

**Comment by alexqyle** [comment]
> The changes look good. Can you please fix the below comment and the ones posted before?

Cool. I will address those comments.

**Comment by alexqyle** [comment]
> @alexqyle can you please fix the code conflict by syncing your branch with the main branch? We should be good to merge after that. Thanks!

@codesome, I have resolved merge conflict. Thanks for all the feedback.

**Comment by alexqyle** [comment]
> @colega rightly pointed out to me that the `Func` suffix on an interface name is confusing. Calling the interface just `PopulateBlock` makes sense. @alexqyle: do you think you can make that change?

Hi @codesome, I can make that change. How about name it `BlockPopulator`? I think make it noun would be better. What do you think?

**Comment by colega** [comment]
I like `BlockPopulator`, I think it's idiomatic Go name for one method interface.
